File: test/StringThing.Npgsql.IntegrationTests/PostgresFixture.cs
using Npgsql;
using Testcontainers.PostgreSql;
using Xunit;

namespace StringThing.Npgsql.IntegrationTests;

public class PostgresFixture : IAsyncLifetime
{
    private readonly PostgreSqlContainer _container = new PostgreSqlBuilder("postgres:17")
        .Build();

    public NpgsqlDataSource DataSource { get; private set; } = null!;

    public async ValueTask InitializeAsync()
    {
        await _container.StartAsync();
        DataSource = NpgsqlDataSource.Create(_container.GetConnectionString());
    }

    public async ValueTask DisposeAsync()
    {
        await DataSource.DisposeAsync();
        await _container.DisposeAsync();
    }
}

File: fsharp/test/StringThing.FSharp.Npgsql.IntegrationTests/PostgresFixture.fs
namespace StringThing.FSharp.Npgsql.IntegrationTests

open System.Threading.Tasks
open Testcontainers.PostgreSql
open Xunit

/// xUnit class fixture that boots a PostgreSQL container once for the test class
/// and tears it down at the end. Image is pinned to `postgres:17` to match the
/// C# integration test suite.
type PostgresFixture() =
    let container = PostgreSqlBuilder().WithImage("postgres:17").Build()
    let mutable connectionString = ""

    member _.ConnectionString = connectionString

    interface IAsyncLifetime with
        member _.InitializeAsync() : ValueTask =
            ValueTask(
                task {
                    do! container.StartAsync()
                    connectionString <- container.GetConnectionString()
                } :> Task)

        member _.DisposeAsync() : ValueTask =
            container.DisposeAsync()
